File: examples/csharp/ProjectorLogCustomer/src/CustomerLogProjectorService.cs
using Angzarr;
using Google.Protobuf.WellKnownTypes;
using Grpc.Core;

namespace Angzarr.Examples.Projector;

public class CustomerLogProjectorService : ProjectorCoordinator.ProjectorCoordinatorBase
{
    private readonly ICustomerLogProjector _projector;

    public CustomerLogProjectorService(ICustomerLogProjector projector)
    {
        _projector = projector;
    }

    public override Task<Empty> Handle(EventBook request, ServerCallContext context)
    {
        _projector.LogEvents(request);
        return Task.FromResult(new Empty());
    }

    public override Task<Projection> HandleSync(EventBook request, ServerCallContext context)
    {
        _projector.LogEvents(request);
        // Log projector doesn't produce a projection
        return Task.FromResult<Projection>(null!);
    }
}

File: examples/java/projector-log-customer/src/main/java/dev/angzarr/examples/projector/CustomerLogProjectorService.java
package dev.angzarr.examples.projector;

import com.google.protobuf.Empty;
import io.grpc.stub.StreamObserver;
import dev.angzarr.EventBook;
import dev.angzarr.Projection;
import dev.angzarr.ProjectorCoordinatorGrpc;

/**
 * gRPC service for customer log projector.
 */
public class CustomerLogProjectorService extends ProjectorCoordinatorGrpc.ProjectorCoordinatorImplBase {

    private final CustomerLogProjector projector;

    public CustomerLogProjectorService(CustomerLogProjector projector) {
        this.projector = projector;
    }

    @Override
    public void handle(EventBook request, StreamObserver<Empty> responseObserver) {
        projector.logEvents(request);
        responseObserver.onNext(Empty.getDefaultInstance());
        responseObserver.onCompleted();
    }

    @Override
    public void handleSync(EventBook request, StreamObserver<Projection> responseObserver) {
        projector.logEvents(request);
        responseObserver.onNext(Projection.getDefaultInstance());
        responseObserver.onCompleted();
    }
}
